### PR TITLE
Use nanoseconds in string to date time conversions

### PR DESCRIPTION
**Pull request by joachimmetz**

### CODE REVIEW

**Review by sydp — CHANGES_REQUESTED** [review]
I left some thoughts for your consideration.

Thanks for adding this, it should help address my other issue on plaso.

File: dfdatetime/cocoa_time.py [hunk 1]
```diff
@@ -78,7 +78,7 @@ def CopyFromDateTimeString(self, time_string):
           YYYY-MM-DD hh:mm:ss.######[+-]##:##
```
[the remark this answers is not in the document]
Comment (joachimmetz) [reply dfdatetime/cocoa_time.py:78]: I'll keep it with 6 given that the docstring can be 3, 6 or 9 and a notation like `#{3,6,9}` might be more confusing

File: dfdatetime/dotnet_datetime.py [hunk 2]
```diff
@@ -94,18 +94,19 @@ def CopyFromDateTimeString(self, time_string):
...
     minutes = date_time_values.get('minutes', 0)
     seconds = date_time_values.get('seconds', 0)
-    microseconds = date_time_values.get('microseconds', 0)
+    nanoseconds = date_time_values.get('nanoseconds', 0)
     time_zone_offset = date_time_values.get('time_zone_offset', 0)
 
     if year > 9999:
       raise ValueError(f'Unsupported year value: {year:d}.')
 
+    nanoseconds, _ = divmod(nanoseconds, 100)
```
[the remark this answers is not in the document]
Comment (joachimmetz) [reply dfdatetime/dotnet_datetime.py:103]: could but divmod is the standard used elsewhere in the codebase

File: dfdatetime/fat_date_time.py [hunk 3]
```diff
@@ -265,20 +265,19 @@ def CopyFromDateTimeString(self, time_string):
...
+
     timestamp = self._GetNumberOfSecondsFromElements(
         year, month, day_of_month, hours, minutes, seconds)
     timestamp -= self._FAT_DATE_TO_POSIX_BASE
     timestamp *= 100
-
-    if microseconds:
-      milliseconds, _ = divmod(microseconds, 10000)
-      timestamp += milliseconds
+    timestamp += milliseconds
```
[the remark this answers is not in the document]
Comment (joachimmetz) [reply dfdatetime/fat_date_time.py:280]: could but divmod is used elsewhere

File: tests/precisions.py [hunk 4]
```diff
@@ -12,12 +12,12 @@
 class DateTimePrecisionHelperTest(unittest.TestCase):
   """Tests for the date time precision helper interface."""
 
-  def testCopyMicrosecondsToFractionOfSecond(self):
-    """Tests the CopyMicrosecondsToFractionOfSecond function."""
+  def testCopNanoosecondsToFractionOfSecond(self):
```
[the remark this answers is not in the document]
Comment (joachimmetz) [reply tests/precisions.py:15]: good catch, changed.

**Review by sydp — APPROVED** [review]
lgtm

**Comment by joachimmetz** [comment]
I'll merge the changes after the next Plaso release

**Pull request merged**